File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    const memoSum = [triangle[0]];
    
    for (let i = 1; i < triangle.length; i++){
        const temp = [];
        for (let j = 0; j <= i; j++){
            let leftSum = 0;
            let rightSum = 0;
            if ((j-1) >= 0){
                leftSum = memoSum[i-1][j-1] + triangle[i][j];
            }
            if (j <= i-1){
                rightSum = memoSum[i-1][j] + triangle[i][j];
            }
            temp.push(Math.max(leftSum, rightSum));
        }
        memoSum.push(temp);
    }
    return Math.max(...memoSum[triangle.length-1]);
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int[][] dp = new int[triangle.length][triangle.length];
        
        for (int i = 0; i < triangle.length; i++){
            int[] temp = new int[i+1];
            for (int j = 0; j < i+1; j++){
                int left = i >0 && j > 0 ? dp[i-1][j-1] : 0;
                int right = i > 0 && j < i ? dp[i-1][j] : 0;
                temp[j] = Math.max(left,right) + triangle[i][j];
            }
            dp[i]= temp;
        }
        
        int answer = 0;
        for (int i : dp[dp.length-1]) {
            answer = Math.max(answer, i);
        }
        return answer;
    }
}